### PR TITLE
BUG: GLM score and hessian

### LINKED ISSUE #4614: BUG: GLM Binomial count, incorrect score, score_obs
It looks like GLM Binomial score and score_obs methods ignore n_trials.

Needs to be confirmed. I am debugging a failure of a new test for penalized GLM Binomial count #4576 and score at IRLS params are very large.

p2 is model estimated with IRLS
res2.model has convergence problems with `bfgs`

```
(Pdb) res2.model.score(p2.params)
array([ 1.82844758, -0.57849861,  1.19638513, -0.32730399])
(Pdb) res2.model.score_obs(p2.params).sum(0)
array([ 1.82844758, -0.57849861,  1.19638513, -0.32730399])
(Pdb) res2.model.score_obs(p2.params).T.dot(res2.model.n_trials)
array([  1.10841891e-11,   1.00175789e-11,   7.95967285e-12,
         7.00894074e-12])
```

maybe the unit tests for the GLM refactoring doesn't have a Binomial count case

### LINKED ISSUE #4616: BUG: GLM families Gamma, inverseGaussian have wrong derivatives
**edit:** see comments at end

The main problem is that loglike and others use the estimated score inside the function, i.e. we concentrate out scale by using a pearson_chi2 by default.
This means for gradient optimization that the derivative with respect to the mean parameters includes the indirect effect of mean parameters on scale and therefore on loglike. The derivative of the loglike as is are not partial derivatives with respect to mean parameters only.
solution: fix scale during optimization

**original description:**

companion to #4614 

I am adding a test to compare score and hessian with numerical derivatives.
The loop over models is in test_glm test_gradient_irls.

Gamma and InverseGaussian have large differences between score and numerical derivative score.
Poisson with sqrt link has some numerical problems, but the power/sqrt link looks fine (based on reading the code)

Other cases than these pass the assert for both score and hessian at rtol=0.05. Tightening the rtol showed failures at around rtol=0.03.

This is on top of the fixes of #4614 that I have currently in my branch.

```
               if not (family_class, link) in [(fam.Poisson, lnk.sqrt),
                                               (fam.Gamma, lnk.log),
                                               (fam.Gamma, lnk.identity),
                                               (fam.Gamma, lnk.inverse_power),
                                               (fam.InverseGaussian, lnk.log),
                                               (fam.InverseGaussian, lnk.identity),
                                               (fam.InverseGaussian, lnk.inverse_power)]:
                   check_score_hessian(rslt_irls)
```

### PR DESCRIPTION
**Pull request by josef-pkt**
fix bugs in GLM score and hessian

first commit are only unit tests to see the results with current code
bugfixes will follow

closes #4616 closes #4614

### CODE REVIEW

**Comment by thequackdaddy** [comment]
This looked good to me. Math makes sense, I think. 

Why did it get closed?

**Comment by josef-pkt** [comment]
I was reading on my ipad, so maybe I accidentally hit the close button.

**Comment by josef-pkt** [comment]
everything was green, but I canceled the ci builds after reopening.

**Comment by josef-pkt** [comment]
this PR sets now scale=1 during gradient optimization.
It covers now all issues raised in #4616.

**Comment by thequackdaddy** [comment]
@josef-pkt This looks really good. Before this, I was never able to use the scipy optimizers without `optim_hessian='eim'`

But now it seems to work well. 

The models I'm running are large-ish. 11.9M records, 53 df model. The `fit` operation takes around 7 minutes for `newton` and `oim`, 6 minutes for `newton` and `eim`, and around 8 for `irls`. 

Curiously, I think the `oim` optimization actually requires slightly fewer iterations... I think 5 vs 4, but I guess each iteration is significantly slower. 

As for `irls`, one on-going issue that I have with it is that it is very memory-intense. On my computer with 64 GB of RAM, `newton` takes a meager 20-25 GB of RAM. `irls` takes over 50... I'm fairly certain that creates a big bottleneck that slows down the irls optimization.

**Comment by josef-pkt** [comment]
another cross check

enabling a disabled discrete  test comparing Probit with GLM, I get difference in hessian at MLE of Probit:

```
hess1 = res1.model.hessian(res1.params)
hess2 = res2.model.hessian(res1.params)

(Pdb) np.abs(np.max(hess1 - hess2))
3.2821745321598428e-11

(Pdb) np.abs(np.max(hess1 / hess2 - 1))
6.8633987382327177e-12
```

So, now they agree, and I should have gotten more suspicious when it didn't agree when I added GLM.hessian.
(Not confirmed numerically: I think in spite of the hessian bug, the hessian with canonical link was still correct, that's why tests didn't fail in typical cases like GLM-Logit, GLM-Poisson or GLM-Gaussian.
The second term for OIM should be zero, and will be zero (+ numerical noise) also with the bug.)

**Comment by josef-pkt** [comment]
This should be finished and ready to merge when green

**Comment by josef-pkt** [comment]
all green, merging

@thequackdaddy  Thanks for checking.

File: statsmodels/discrete/tests/test_sandwich_cov.py [hunk 1]
```diff
@@ -479,13 +479,9 @@ def test_basic(self):
...
         assert_equal(res2.cov_type, self.cov_type)
 
-        assert_allclose(res1.params, res2.params, rtol=1e-13)
-        # bug TODO res1.scale missing ?  in Gaussian/OLS
-        assert_allclose(res1.bse, res2.bse, rtol=1e-13)
-#         if not self.cov_type == 'nonrobust':
-#             assert_allclose(res1.bse * res1.scale, res2.bse, rtol=1e-13)
-#         else:
-#             assert_allclose(res1.bse, res2.bse, rtol=1e-13)
+        rtol = getattr(res1, 'rtol', 1e-13)
```
Comment (jbrockmendel) [inline statsmodels/discrete/tests/test_sandwich_cov.py]: should this be getting `cls.rtol`?  That is the attribute that is set down on L514.

Comment (josef-pkt) [reply statsmodels/discrete/tests/test_sandwich_cov.py]: Yes, that's a mistake.
However, why do the tests not fail if rtol is 1e-13?

Essentially this means that GLM and the corresponding discrete model agree on very high precision with the same optimizer newton. 1e-13 should be much stricter than convergence tolerance.

**Pull request merged**